bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,1076 @@
+     * hub will create a new one. Note: this requires an HTTP 1.1 client call.
+     *
+     * @param oldApplicationKey existing application key if any i.e. may be empty.
+     * @return the existing or a newly created application key.
+     * @throws ApiException if there was a communications error.
+     * @throws HttpUnauthorizedException if the registration failed.
+     */
+    public String registerApplicationKey(@Nullable String oldApplicationKey)
+            throws ApiException, HttpUnauthorizedException {
+        logger.debug("registerApplicationKey()");
+        String json = jsonParser.toJson((Objects.isNull(oldApplicationKey) || oldApplicationKey.isEmpty())
+                ? new CreateUserRequest(APPLICATION_ID)
+                : new CreateUserRequest(oldApplicationKey, APPLICATION_ID));
+        Request httpRequest = httpClient.newRequest(registrationUrl).method(HttpMethod.POST)
+                .timeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
+                .content(new StringContentProvider(json), MediaType.APPLICATION_JSON);
+        ContentResponse contentResponse;
+        try {
+            logger.trace("POST {} HTTP/1.1 >> {}", registrationUrl, json);
+            contentResponse = httpRequest.send();
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            throw new ApiException("HTTP processing error", e);
+        }
+        int httpStatus = contentResponse.getStatus();
+        json = contentResponse.getContentAsString().trim();
+        logger.trace("HTTP/1.1 {} {} << {}", httpStatus, contentResponse.getReason(), json);
+        if (httpStatus != HttpStatus.OK_200) {
+            throw new ApiException("HTTP bad response");
+        }
+        try {
+            List<SuccessResponse> entries = jsonParser.fromJson(json, SuccessResponse.GSON_TYPE);
+            if (Objects.nonNull(entries) && !entries.isEmpty()) {
+                SuccessResponse response = entries.get(0);
+                Map<String, Object> responseSuccess = response.success;
+                if (Objects.nonNull(responseSuccess)) {
+                    String newApplicationKey = (String) responseSuccess.get("username");
+                    if (Objects.nonNull(newApplicationKey)) {
+                        return newApplicationKey;
+                    }
+                }
+            }
+        } catch (JsonParseException e) {
+            // fall through
+        }
+        throw new HttpUnauthorizedException("Application key registration failed");
+    }
+
+    public void setRestartScheduled() {
+        restartScheduled = true;
+        close();
+    }
+
+    /**
+     * Sleep the caller during any period when the connection is marked as restarting.
+     * Force time out after 10 seconds.
+     */
+    private void sleepDuringRestart() {
+        int iteration = 0;
+        while (restarting) {
+            try {
+                Thread.sleep(500);
```

Comment:
Please see https://github.com/openhab/openhab-addons/pull/12342#discussion_r1237377151. I think this comment also applies here, but perhaps @fwolter can advise?

Reply:
In case `sleepDuringRestart()` is run in a framework's scheduler (without creating your own, which should be avoided), you could schedule a periodic task checking wether restarting is true and if so, execute whatever comes after `sleepDuringRestart()`.

Reply:
Sorry but I cannot do what you ask.

The overall purpose of the outer method that calls this inner method is to make a BLOCKING HTTP GET or PUT request to the Hue bridge server. And as you know such GET's and PUT's do (even in normal circumstances) take time to execute. Therefore in any case the outer method is called from a thread pool, and definitely not directly from the OH core. Under special circumstances, the sleep loop within `sleepDuringRestart()` may extend the blocking duration of the HTTP call. The special circumstance is the case when a GET/PUT call is made during the interval when the underlying HTTP2 session is being renewed. It will simply adds an extra delay to the HTTP2 GET/PUT stream call until the underlying HTTP2 session opening process is complete. In other words it just delays the blocking duration of the overall call.

To summarise: the good news is that the task is not called directly from OH core, so it cannot block OH.

Note: that in this case I could (indeed I shall) replace the Thread.sleep() with a wait for a mutex on the underlying HTTP2 session. That way, the internal logic would become clearer.

Reply:
> The special circumstance is the case when a GET/PUT call is made during the interval when the underlying HTTP2 session is being renewed.

Can you give an example about the GET/PUT request? Is this request triggered by the user or by some periodic tasks by the binding? Why does the request need to be blocking? Do you need to return a synchronous result?

Reply:
@fwolter two things..

1. **GET** -- As a general rule the binding relies on SSE events to provide channel state. However SSE only reports the 'delta' from prior state, so when the thing initially starts we do a series of GET's to fetch the current actual state. Also from time to time it does a full GET series to resynchronize the current actual state in case any SSE event may have been dropped. These GETs have to wait until the respective data is returned. They are made from an OH core scheduled task so they should not block the core. And they are made with a socket request timeout, so they cannot block indefinitely.

2. **PUT** -- These are used by the binding 'handleCommand()' method, when the system wants to change the state of something. The call is made on whatever thread of OH called 'handleCommand()' . The calls have to wait for the HTTP response to know if the call succeeded or failed. They are made with a socket request timeout, so they cannot block indefinitely.

The Hue bridge server HTTP2 implementation drops the HTTP2 session with a HTTP2 GO_AWAY message every hour or two. This is a 'normal' error from the bridge server, probably for house keeping purposes. If the bridge server has dropped the session with a GO_AWAY then it won't accept a a reconnect attempt until about 2..3 seconds later. Therefore if by chance the binding wants to make a GET or a PUT at the same time as the HTTP2 session is in the GO-AWAY / reconnection phase, then the respective GET/PUT is put on sleep until that action has completed. In that case the worst case delay is the session restart sleep time (maximum 5 seconds) plus the socket request timeout.

Reply:
So, from time to time, switching a light takes 5s? Are you sure this is the intended behavior by the API delevopers? Is there any chance you can proactively renew the session before it expires?

Reply:
> from time to time, switching a light takes 5s

Yes.

> this is the intended behavior by the API delevopers

No.

---

I did not test the extreme limits of this edge case. What I **have** observed is the following..
1. From "time to time" the Hue bridge server issues a GO_AWAY to force close the session. I did not measure "time to time" exactly but it is certainly many hours if not one day. So yes indeed doing a pre-emptive teardown and restart might avoid the forced GO_AWAY cases. (But I would advise to keep the code that handles hard GO_AWAY cases anyway).

2. After doing a session teardown the Hue bridge server will not accept new connection attempts "for some time". I did not measure "for some time" exactly, and I did not test if it is longer in case of a server forced GO_AWAY than in case of a client initiated teardown. I rather randomly chose 5 seconds which certainly safe enough to work consistently.

So I propose the following next steps..
- I shall let the binding run until I have encountered the next two forced GO_AWAYs -- this will allow us to determine the meaning of "time to time".
- After that I shall do some increasing strict teardown and restart tests to establish the lower limit of the dead period -- this will allow us to determine the meaning of "for some time".

Note: these tests could take up to 2 days, so your M4 deadline might be difficult to achieve.

Reply:
Sounds like a reasonable approach! I'm good to merge.

Reply:
A quick update:

Since I started work on this PR there have been five Hue bridge firmware updates. When I wrote my original JavaDoc I had obviously observed that -- with the then firmware -- the "time to time" GO-AWAY interval was nine or ten hours. However as of my current test run -- with the May 25th 2023 firmware -- I have not yet observed a any GO_AWAY cases even after 22 hours continuous uptime (and counting). So it may be that Signify has fixed the issue in the meantime anyway. The firmware release notes always just say "general quality and stability improvements, blah blah" so we can't know for sure. But anyway I will let my test run for a couple more hours to see if there is a GO_AWAY around the 24 hours mark.

BTW please note that there was yet another new firmware released two days ago, but I have not yet installed that, since I want to let the current test run complete uninterrupted. AFAICT Signify is currently pushing new firmware almost every month; I think this is mainly in preparation for their (long overdue) official Matter release. (Currently Hue on Matter is only available via their limited beta program).

Reply:
@fwolter / @jlaur I need your advice: the results of my tests are as follows..

- The session has been up for 25+ hours on my operative OH system without any GO_AWAY so far. I will continue monitoring until I get the first one (if any) and report back.

- If the binding manually tears down the session (on the client side) then the bridge server permits reconnection within 100 milliseconds or less.

- Due to lack of any any GO_AWAY from the bridge server side I could not check it today; however my memory is that if the server had sent a GO_AWAY then it refuses a new reconnect from the same client for "some seconds". But as I said, I could not establish how long, although we do know that the current 5 seconds is 'enough' (i.e. more than enough).

Therefore we have the following options to proceed..

1. Keep the code as it is. This means that if the bridge server would send a GO_AWAY, which may happen on a timescale of days rather than hours, then the binding will reconnect automatically, whereby any command received during the reconnect window might be delayed up to 5 seconds. So the risk duty cycle (5 seconds in 25+ hours) could cause delayed execution of 0.006% of potential commands by an average of 2.5 seconds. IMHO this is an acceptable risk.

2. Implement, as you suggest, a periodic pre-emptive client side teardown and reconnect schedule. (At what interval?) This could reduce the above-mentioned 0.006% risk duty cycle, albeit NOT to zero since one should anyway keep some safety window). Furthermore at some point if the periodic pre-emptive interval becomes too short, then the duty cycle would eventually exceed that of case 1. And finally this could potentially introduce other unforeseen errors, since I have never actually run such scenario over a long period of time.

My personal suggestion is scenario 1. "better the devil you know". And we can release it as is, and fix it later if we encounter adverse feedback from users. However if you want 2. then please say so (and if so please suggest your parameters for the pre-emptive restart period, and the teardown / restart recycle interval).

Reply:
As already said, I'm okay with merging the current code.

If it turns out the GO_AWAY never happens under normal conditions, I'd go for removing the re-connect within the HTTP command handling code to make it only blocking for *one* HTTP connection. If it turns out the session isn't alive anymore during execution of such an HTTP command, I'd find it okay if this specific HTTP command fails, and the Thing handler logic takes care the session is re-established asynchronously. Assuming that this case is exceptional and doesn't happen normally. Maybe the Thing handler polls the API periodically anyway and therefore checks regularly if the session is still alive without user interaction.

Reply:
> Maybe the Thing handler polls the API periodically anyway and therefore checks regularly if the session is still alive without user interaction

I confirm it does that already anyway; it sends HTTP2 ping frames.

Reply:
> If it turns out the GO_AWAY never happens under normal conditions

Curiously I got three such GO_AWAYs last night. The first occurred 29 hours after OH start up and there were another two within a 15 minutes window. But then again since that time it has been running steadily for the last 23+ hours. (And I could identify no peculiar pattern of activity on the bridge that might indicate why the GO_AWAYs came at that time). So I think it is good to keep the GO_AWAY handler code as it is. At least until I or another user can figure out if there is any (avoidable) pattern to these cases.